[Dialogs] Allow setting text alignment for alert message text

galiak11 opened the pull request:
# Description

Using NSTextAlignment to customize alignment of the Material Alert's message.

## Example
Trailing aligned message with a leading aligned message:

![Dialogs - message alignment](https://user-images.githubusercontent.com/2329102/76005569-35198300-5ed9-11ea-9d4d-a2d75469a477.png)

## Issue
Closes #5602 

b/118846596, cl/298994401

material-automation commented:
bazel detected changes to the following targets:

```
//components/Dialogs:snapshot_tests
//components/Dialogs:snapshot_tests_test_bundle
//components/Dialogs:unit_tests
//components/Dialogs:unit_tests_IPAD_PRO_12_9_IN_9_3
//components/Dialogs:unit_tests_IPAD_PRO_12_9_IN_9_3_test_bundle
//components/Dialogs:unit_tests_IPHONE_7_PLUS_IN_10_3
//components/Dialogs:unit_tests_IPHONE_7_PLUS_IN_10_3_test_bundle
//components/Dialogs:unit_tests_environment
//components/Dialogs:unit_tests_environment_test_bundle
```

<!-- identifier: affected-targets -->

material-automation commented:
## API diff detected the following changes

### Dialogs

#### MDCAlertControllerView

*new* property: `messageAlignment` in `MDCAlertControllerView`

*modified* class: `MDCAlertControllerView`

| Type of change: | Declaration |
|---|---|
| From: | `@interface MDCAlertControllerView : UIView  @property(nonatomic, strong, nullable) UIFont *titleFont UI_APPEARANCE_SELECTOR; @property(nonatomic, strong, nullable) UIColor *titleColor UI_APPEARANCE_SELECTOR;  @property(nonatomic, assign) NSTextAlignment titleAlignment; @property(nonatomic, strong, nullable) UIImage *titleIcon; @property(nonatomic, strong, nullable) UIColor *titleIconTintColor;  /**  An optional custom icon view above the title of the alert.   @note This property is intended to be used to provide a custom implementation of the title icon  view. If the intention is to just display a `UIImage`, use `setTitleIcon:` API instead. If  'titleIcon' is set, 'titleIconView' is ignored.  */ @property(nonatomic, strong, nullable) UIView *titleIconView;  @property(nonatomic, strong, nullable) UIFont *messageFont UI_APPEARANCE_SELECTOR; @property(nonatomic, strong, nullable) UIColor *messageColor UI_APPEARANCE_SELECTOR;  // b/117717380: Will be deprecated (x3) @property(nonatomic, strong, nullable) UIFont *buttonFont UI_APPEARANCE_SELECTOR; @property(nonatomic, strong, nullable) UIColor *buttonColor UI_APPEARANCE_SELECTOR; @property(nonatomic, strong, nullable) UIColor *buttonInkColor UI_APPEARANCE_SELECTOR;  @property(nonatomic, assign) CGFloat cornerRadius;  /*  Indicates whether the view's contents should automatically update their font when the device’s  UIContentSizeCategory changes.   This property is modeled after the adjustsFontForContentSizeCategory property in the  UIContentSizeCategoryAdjusting protocol added by Apple in iOS 10.   Default value is NO.  */ @property(nonatomic, readwrite, setter=mdc_setAdjustsFontForContentSizeCategory:)     BOOL mdc_adjustsFontForContentSizeCategory UI_APPEARANCE_SELECTOR;  /**  By setting this property to @c YES, the Ripple component will be used instead of Ink  to display visual feedback to the user.   @note This property will eventually be enabled by default, deprecated, and then deleted as part  of our migration to Ripple. Learn more at  https://github.com/material-components/material-components-ios/tree/develop/components/Ink#migration-guide-ink-to-ripple   Defaults to NO.  */ @property(nonatomic, assign) BOOL enableRippleBehavior;  @end` |
| To: | `@interface MDCAlertControllerView : UIView  @property(nonatomic, strong, nullable) UIFont *titleFont UI_APPEARANCE_SELECTOR; @property(nonatomic, strong, nullable) UIColor *titleColor UI_APPEARANCE_SELECTOR;  @property(nonatomic, strong, nullable) UIImage *titleIcon; @property(nonatomic, strong, nullable) UIColor *titleIconTintColor; @property(nonatomic, assign) NSTextAlignment titleAlignment; @property(nonatomic, assign) NSTextAlignment messageAlignment;  /**  An optional custom icon view above the title of the alert.   @note This property is intended to be used to provide a custom implementation of the title icon  view. If the intention is to just display a `UIImage`, use `setTitleIcon:` API instead. If  'titleIcon' is set, 'titleIconView' is ignored.  */ @property(nonatomic, strong, nullable) UIView *titleIconView;  @property(nonatomic, strong, nullable) UIFont *messageFont UI_APPEARANCE_SELECTOR; @property(nonatomic, strong, nullable) UIColor *messageColor UI_APPEARANCE_SELECTOR;  // b/117717380: Will be deprecated (x3) @property(nonatomic, strong, nullable) UIFont *buttonFont UI_APPEARANCE_SELECTOR; @property(nonatomic, strong, nullable) UIColor *buttonColor UI_APPEARANCE_SELECTOR; @property(nonatomic, strong, nullable) UIColor *buttonInkColor UI_APPEARANCE_SELECTOR;  @property(nonatomic, assign) CGFloat cornerRadius;  /*  Indicates whether the view's contents should automatically update their font when the device’s  UIContentSizeCategory changes.   This property is modeled after the adjustsFontForContentSizeCategory property in the  UIContentSizeCategoryAdjusting protocol added by Apple in iOS 10.   Default value is NO.  */ @property(nonatomic, readwrite, setter=mdc_setAdjustsFontForContentSizeCategory:)     BOOL mdc_adjustsFontForContentSizeCategory UI_APPEARANCE_SELECTOR;  /**  By setting this property to @c YES, the Ripple component will be used instead of Ink  to display visual feedback to the user.   @note This property will eventually be enabled by default, deprecated, and then deleted as part  of our migration to Ripple. Learn more at  https://github.com/material-components/material-components-ios/tree/develop/components/Ink#migration-guide-ink-to-ripple   Defaults to NO.  */ @property(nonatomic, assign) BOOL enableRippleBehavior;  @end` |

#### MDCAlertController

*new* property: `messageAlignment` in `MDCAlertController`

<!-- identifier: apidiff -->

jverkoey commented:
Please associate the PR with the GitHub issue for tracking purposes by putting something like:

`Closes <link to GitHub Issue>`

in the PR description.

jverkoey commented on components/Dialogs/src/MDCAlertController.m:
```diff
@@ -329,6 +330,14 @@ - (void)setTitleAlignment:(NSTextAlignment)titleAlignment {
   }
 }
 
+// todo: test alignment getter - after changing the alert.view.messageAlignment
```
Please file an Issue to track this TODO and link to it here in the form `TODO(<link to Issue>): ...`.

galiak11 replied:
Removed the comment.

I tested this, and it seems all Dialogs getters return incorrect values after a different value is set directly on the view. For example:
```
MDCAlertController *alertview = (MDCAlertController *)alert.view;
alertview.<getter> = <value>; 
print (alert.<getter>);   // prints original value before the change.
```
To remain consistent with other getters, I repeated this behavior in the new getter.

jverkoey reviewed (commented):
> Trailing aligned message with a leading aligned message:

Do you mean to say "An example with a trailing-aligned message" ?

jverkoey reviewed (approved):
LGTM, but please add a snapshot test for the new API prior to landing this PR.

material-automation commented on components/Dialogs/tests/snapshot/MDCAlertControllerConfigurationsTests.m:
```diff
@@ -41,7 +41,7 @@ - (void)setUp {
   // Uncomment below to recreate all the goldens (or add the following line to the specific
   // test you wish to recreate the golden for).
-  //  self.recordMode = YES;
+    self.recordMode = YES;
```
clang-format suggested the following change:


```
-    self.recordMode = YES;
+  self.recordMode = YES;
```

<!-- identifier: clang-format -->
<!-- File: components/Dialogs/tests/snapshot/MDCAlertControllerConfigurationsTests.m -->

material-automation commented on components/Dialogs/tests/snapshot/MDCAlertControllerConfigurationsTests.m:
```diff
@@ -375,4 +375,116 @@ - (void)testAlertHasTitleAndTitleIconAndMessageAndAccessoryViewInRTL {
+  [self generateSizedSnapshotAndVerifyForView:self.alertController.view];
+}
+
+
```
clang-format suggested the following change:


```
-
```

<!-- identifier: clang-format -->
<!-- File: components/Dialogs/tests/snapshot/MDCAlertControllerConfigurationsTests.m -->

material-automation commented:
clang-format recommended changes.

Consider installing [git-clang-format](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/git-clang-format) and running the following command:

```bash
git clang-format $(git merge-base origin/develop HEAD)
```

<!-- identifier: clang-format -->

material-automation commented on components/Dialogs/src/MDCAlertController.h:
```diff
@@ -113,12 +113,15 @@
+/** The color applied to the message of the Alert Controller.*/
 @property(nonatomic, strong, nullable) UIColor *messageColor;
 
+/** The alignment applied to the message of the Alert Controller. Default to NSTextAlignmentNatural. */
```
clang-format suggested the following change:


```
-/** The alignment applied to the message of the Alert Controller. Default to NSTextAlignmentNatural. */
+/** The alignment applied to the message of the Alert Controller. Default to NSTextAlignmentNatural.
+ */
```

<!-- identifier: clang-format -->
<!-- File: components/Dialogs/src/MDCAlertController.h -->

material-automation commented on components/Dialogs/src/MDCAlertController.h:
```diff
@@ -98,27 +98,30 @@
 @property(nonatomic, strong, nullable) UIColor *titleColor;
 
-/** The alignment applied to the title of the Alert Controller.*/
+/** The alignment applied to the title of the Alert Controller. Default to NSTextAlignmentNatural. */
```
clang-format suggested the following change:


```
-/** The alignment applied to the title of the Alert Controller. Default to NSTextAlignmentNatural. */
+/** The alignment applied to the title of the Alert Controller. Default to NSTextAlignmentNatural.
+ */
```

<!-- identifier: clang-format -->
<!-- File: components/Dialogs/src/MDCAlertController.h -->

jverkoey commented on components/Dialogs/src/MDCAlertController.h:
```diff
@@ -98,27 +98,30 @@
 @property(nonatomic, strong, nullable) UIFont *messageFont;
 
-/** The color applied to the message of Alert Controller.*/
+/** The color applied to the message of the Alert Controller. */
```
Please move the unrelated text changes here out to a separate cleanup PR.

jverkoey reviewed (commented):
testMessageRightAlignmentIsRightInRTL_11_2@2x.png looks like it may have uncovered a bug. Please file an issue for it accordingly.

galiak11 committed:
[Dialogs] Allow setting text alignment for alert message text

Using NSTextAlignment to customize alignment of the alert's message.

b/118846596, cl/298994401
PiperOrigin-RevId: 298994401

galiak11 committed:
Adding message alignment snapshot tests

galiak11 commented:
These text changes were added in response to review feedback in corresponding CL.

jverkoey commented:
This was rolled back in https://github.com/material-components/material-components-ios/pull/9905

https://github.com/material-components/material-components-ios/issues/9909 to track rolling forward.

jverkoey commented:
Rolled forward in the release. The internal testing issue appears to be due to timeouts and not these changes.
